File: data_structures/bloom_filter/cpp/bloom_filter.cpp
#include "bloom_filter.h"

BloomFilter::BloomFilter(std::size_t m, int k) : m_(m), k_(k), bits_(m, false) {}

std::size_t BloomFilter::hash(int value, int seed) const {
    unsigned int h = static_cast<unsigned int>(value) ^ static_cast<unsigned int>(seed * 0x9e3779b9);
    h ^= h >> 16;
    h *= 0x85ebca6b;
    h ^= h >> 13;
    h *= 0xc2b2ae35;
    h ^= h >> 16;
    return h % m_;
}

void BloomFilter::insert(int value) {
    for (int i = 0; i < k_; ++i)
        bits_[hash(value, i)] = true;
}

bool BloomFilter::contains(int value) const {
    for (int i = 0; i < k_; ++i)
        if (!bits_[hash(value, i)]) return false;
    return true;
}

File: data_structures/bloom_filter/python/bloom_filter.py
"""ブルームフィルタ

操作:
  insert(v)   : O(k)
  contains(v) : O(k) — 偽陽性あり、偽陰性なし

k 個の独立ハッシュ関数で m ビットのビット配列を使用。
"""


from typing import Generic, TypeVar

T = TypeVar("T")


class BloomFilter(Generic[T]):
    def __init__(self, m: int, k: int) -> None:
        self._m = m
        self._k = k
        self._bits = bytearray(m)

    def _hash(self, value: T, seed: int) -> int:
        h = hash(value) ^ (seed * 0x9E3779B9)
        h &= 0xFFFFFFFF
        h ^= h >> 16
        h = (h * 0x85EBCA6B) & 0xFFFFFFFF
        h ^= h >> 13
        h = (h * 0xC2B2AE35) & 0xFFFFFFFF
        h ^= h >> 16
        return h % self._m

    def insert(self, value: T) -> None:
        for i in range(self._k):
            self._bits[self._hash(value, i)] = 1

    def contains(self, value: T) -> bool:
        return all(self._bits[self._hash(value, i)] for i in range(self._k))
